File: video_engine/src/animation/keyframe_engine.cpp
#include "keyframe_engine.h"
#include <algorithm>
#include <cmath>

namespace VideoEngine {

float KeyframeEngine::evaluate(const std::vector<Keyframe>& keyframes, double currentTimeUs) {
    if (keyframes.empty()) return 0.0f;
    if (keyframes.size() == 1 || currentTimeUs <= keyframes.front().timeUs)
        return keyframes.front().value;
    if (currentTimeUs >= keyframes.back().timeUs)
        return keyframes.back().value;

    auto it = std::lower_bound(keyframes.begin(), keyframes.end(), currentTimeUs,
        [](const Keyframe& k, double t) { return k.timeUs < t; });

    const Keyframe& next = *it;
    const Keyframe& prev = *(--it);

    float t = (float)((currentTimeUs - prev.timeUs) / (next.timeUs - prev.timeUs));

    if (next.easing == EasingType::BEZIER) {
        return interpolateBezier(prev.value, next.value, t, next.cp1x, next.cp1y, next.cp2x, next.cp2y);
    } else if (next.easing == EasingType::HOLD) {
        return prev.value;
    }

    return prev.value + (next.value - prev.value) * t;
}

float KeyframeEngine::interpolateBezier(float start, float end, float t, float x1, float y1, float x2, float y2) {
    float cx = 3.0f * x1;
    float bx = 3.0f * (x2 - x1) - cx;
    float ax = 1.0f - cx - bx;

    auto sampleCurveX = [&](float t) { return ((ax * t + bx) * t + cx) * t; };

    float t0 = t;
    for (int i = 0; i < 8; i++) {
        float x2_val = sampleCurveX(t0) - t;
        if (fabs(x2_val) < 1e-6) break;
        float d2 = (3.0f * ax * t0 + 2.0f * bx) * t0 + cx;
        if (fabs(d2) < 1e-6) break;
        t0 = t0 - x2_val / d2;
    }

    float cy = 3.0f * y1;
    float by = 3.0f * (y2 - y1) - cy;
    float ay = 1.0f - cy - by;
    return start + (end - start) * ((ay * t0 + by) * t0 + cy) * t0;
}

}

File: flutter_app/lib/core/engine/keyframe_engine.dart
import 'dart:math';

enum EasingType { linear, bezier, hold }

class Keyframe {
  final double timeUs;
  final double value;
  final EasingType easing;
  final double cp1x, cp1y, cp2x, cp2y;

  Keyframe({
    required this.timeUs,
    required this.value,
    this.easing = EasingType.linear,
    this.cp1x = 0.42,
    this.cp1y = 0.0,
    this.cp2x = 0.58,
    this.cp2y = 1.0,
  });
}

class KeyframeEngine {
  double evaluate(List<Keyframe> keyframes, double currentTimeUs) {
    if (keyframes.isEmpty) return 0.0;
    if (keyframes.length == 1 || currentTimeUs <= keyframes.first.timeUs) {
      return keyframes.first.value;
    }
    if (currentTimeUs >= keyframes.last.timeUs) {
      return keyframes.last.value;
    }

    // Find the current segment
    int nextIdx = keyframes.indexWhere((k) => k.timeUs > currentTimeUs);
    final prev = keyframes[nextIdx - 1];
    final next = keyframes[nextIdx];

    double t = (currentTimeUs - prev.timeUs) / (next.timeUs - prev.timeUs);

    if (next.easing == EasingType.bezier) {
      return _interpolateBezier(prev.value, next.value, t, next.cp1x, next.cp1y, next.cp2x, next.cp2y);
    } else if (next.easing == EasingType.hold) {
      return prev.value;
    }

    return prev.value + (next.value - prev.value) * t;
  }

  double _interpolateBezier(double start, double end, double t, double x1, double y1, double x2, double y2) {
    // Cubic Bezier Ease-in-out calculation
    // Simplified for Dart; production uses Newton-Raphson as in C++
    double factor = pow(1 - t, 3) * 0 + 3 * pow(1 - t, 2) * t * x1 + 3 * (1 - t) * pow(t, 2) * x2 + pow(t, 3) * 1;
    return start + (end - start) * factor;
  }
}
